src/storage/hummock_trace/src/replay/runner.rs

```diff
@@ -0,0 +1,261 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+use std::time::Instant;
+
+use crate::error::Result;
+use crate::read::TraceReader;
+use crate::{Operation, Replayable, WorkerScheduler};
+
+pub struct HummockReplay<R: TraceReader> {
+    reader: R,
+    replay: Arc<Box<dyn Replayable>>,
+}
+
+impl<R: TraceReader> HummockReplay<R> {
+    pub fn new(reader: R, replay: Box<dyn Replayable>) -> Self {
+        Self {
+            reader,
+            replay: Arc::new(replay),
+        }
+    }
+
+    pub async fn run(&mut self) -> Result<()> {
+        let time = Instant::now();
+        let mut worker_scheduler = WorkerScheduler::new();
+        let mut total_ops: u64 = 0;
+
+        while let Ok(r) = self.reader.read() {
+            match r.op() {
+                Operation::Result(trace_result) => {
+                    worker_scheduler.send_result(&r, trace_result.to_owned());
+                }
+                Operation::Finish => {
+                    worker_scheduler.wait_finish(&r).await;
+                }
+                _ => {
+                    worker_scheduler.schedule(r, self.replay.clone());
```

Comment:
We can store an instance of `self.replay` in the worker scheduler so that we don't need a copy here.